Shopify/ruby-lsp
# Avoid making semantic tokens for special ruby methods

### Motivation

The default syntax highlighting (by [textmate](https://github.com/textmate/ruby.tmbundle)) for Ruby accounts for Ruby-specific language methods, like `require_relative`, `include`, `attr_reader`. 

Currently, our semantic highlighting considers those special methods as a regular method. 

However, we want to keep feature parity and would like to maintain those special methods as a different highlight than a regular method.

### Implementation

There is a list of special methods listed by textscope: [gemfile methods](https://github.com/textmate/ruby.tmbundle/blob/efcb8941c701343f1b2e9fb105c678152fea6892/Syntaxes/Gemfile.tmLanguage#L21), [requires](https://github.com/textmate/ruby.tmbundle/blob/efcb8941c701343f1b2e9fb105c678152fea6892/Syntaxes/Ruby.plist#L329), and [ruby language specific methods](https://github.com/textmate/ruby.tmbundle/blob/efcb8941c701343f1b2e9fb105c678152fea6892/Syntaxes/Ruby.plist#L323). I've exported this list into our request class. 

Essentially, for all the types of method call nodes, avoid adding a new semantic token if the method value is a special method.  

### Automated Tests

Covered in the new test case. 

### Manual Tests

As usual, you can turn on VS Code's developer Inspect Editor Tokens and Scopes node to inspect those special methods to see that there isn't a semantic token we've added. Plus the highlighting for these special methods should not be the same as a regular method call.

#### Before: Notice same highlight for special methods and regular method call
<img width="419" alt="Screen Shot 2022-06-21 at 12 06 36 PM" src="https://user-images.githubusercontent.com/25471753/174785298-fe1ee5be-4d79-4d84-baca-8a2ec6228a33.png">

#### After: Much more obvious which method calls are special ones
<img width="451" alt="Screen Shot 2022-06-21 at 12 08 26 PM" src="https://user-images.githubusercontent.com/25471753/174785608-fc4915c8-bbcc-42ac-80a3-452a2e071927.png">
